calculate bignum sign without multiply

gstrauss opened the pull request:
## Description

calculate bignum sign without multiply

This PR includes commits from PRs #6585 #6587 #6588.  The last commit in this PR calculates bignum sign without using multiply.

## Gatekeeper checklist

- [ ] **changelog** provided, or not required
- [ ] **backport** done, or not required
- [ ] **tests** provided, or not required

tom-cosgrove-arm commented on library/bignum.c:
```diff
@@ -1372,16 +1365,19 @@ int mbedtls_mpi_div_mpi( mbedtls_mpi *Q, mbedtls_mpi *R, const mbedtls_mpi *A,
+        T2.s = B->s; /* save sign value in case outputs alias inputs */
         MBEDTLS_MPI_CHK( mbedtls_mpi_copy( Q, &Z ) );
-        Q->s = A->s * B->s;
+        Q->s = T1.s ^ T2.s ? -1 : 1;
```
isn't `1 + ( T1.s ^ T2.s )` better?

gstrauss replied:
I changed it to `( T1.s ^ T2.s ) | 1`
:rocket:

tom-cosgrove-arm commented:
Not sure why this PR requires the preceding PRs - it's quite a simple change, that is going to be faster to review than the others

gstrauss commented:
> Not sure why this PR requires the preceding PRs - it's quite a simple change, that is going to be faster to review than the others

Since you prefer this one first, I have rebased and removed the commits from the other PRs.
I'll fix the merge conflicts in one of the other PRs once this is accepted.

gstrauss commented:
This PR no longer needs the label "needs-preceding-pr".

gstrauss commented:
This PR will generate merge conflicts with #6588 
I have added tests to all three of the following small, related PRs, and would prefer if they were reviewed before this.  Thank you.
#6588 mbedtls_mpi_mul_mpi allow output to alias inputs using single temporary mbedtls_mpi
#6585 mbedtls_mpi_div_mpi allow outputs to alias inputs
#6587 mbedtls_mpi_exp_mod allow output to alias inputs

tom-cosgrove-arm commented:
Looks like the PR merge TLS Testing all_u16-test_psa_crypto_config_accel_hash_use_psa failure was due to timeout on the system

```
docker run -u $(id -u):$(id -g) -e MAKEFLAGS --rm --entrypoint /var/lib/build/steps.sh -w /var/lib/build -v `pwd`/src:/var/lib/build --cap-add SYS_PTRACE 666618195821.dkr.ecr.eu-west-1.amazonaws.com/jenkins-mbedtls:ubuntu-16.04-c1f96fa11a6066680482ab9c0ed3ed84daa47e99
— Shell Script
1h 0m 35s
```

gilles-peskine-arm commented on library/bignum.c:
```diff
@@ -983,7 +983,7 @@ int mbedtls_mpi_add_mpi( mbedtls_mpi *X, const mbedtls_mpi *A, const mbedtls_mpi
 
     s = A->s;
-    if( A->s * B->s < 0 )
+    if( A->s ^ B->s )
```
That's a fancy way of writing `if( A->s != B->s )`, with the added twist that the combination of bitwise operators and signed integers makes the reader pause and think whether it might create a trap representation. (It won't, because the only allowed values are 1 and -1.)

gstrauss replied:
...and how excessive is `A->s * B->s > 0` to do the same?
:rocket:

gilles-peskine-arm commented on library/bignum.c:
```diff
@@ -1018,7 +1018,7 @@ int mbedtls_mpi_sub_mpi( mbedtls_mpi *X, const mbedtls_mpi *A, const mbedtls_mpi
 
     s = A->s;
-    if( A->s * B->s > 0 )
+    if( !( A->s ^ B->s ) )
```
As above, this is a fancy way of writing `A->s == B->s`.

gstrauss replied:
As above: ...and how excessive is `A->s * B->s > 0` to do the same?
:rocket:

gilles-peskine-arm reviewed (changes requested):
It's not clear to me that this change is really useful. Do the use of a multiplication instruction and the potential branching on `result_is_zero` result in measurably less efficient or larger code?

If this is justified by a benchmark or a code size measurement, please share.

In any case, please optimize for readable code first.

gstrauss commented on library/bignum.c:
Do you prefer this?
`X->s = ( ( A->s != B->s ) & ( i != 0 ) & ( j != 0 ) ) ? -1 : 1;`

There is  a comment *directly* above this line of code about constant time operations and the need to check if the result is 0.  This code is better to hide that check than the timing of the previous branching.

What kind of benchmark would you like to demonstrate that multiplication is many times the number of clock cycles than XOR or BNE.  Would a timing loop of 1 million operations suffice?

tom-cosgrove-arm replied:
I don't see the benefit of removing the `result_is_zero` variable, as it doesn't remove any of the comparisons. The compiler might well optimise out the variable, but doing it in the code makes the code harder to read

gstrauss replied:
@tom-cosgrove-arm, the previous code uses branches to set `result_is_zero`.  That may leak timing information.
Would you prefer two lines of code:
```
i = ( i != 0 ) & ( j != 0 ); /* result is not zero */
X->s = ( ( A->s != B->s ) & i ) ? -1 : 1;
```

gilles-peskine-arm replied:
The current code still uses branches, they just have different timing (if not reordered by the compiler). The timing of the loops to calculate `i` and `j` leak the same information anyway — it's the number of iterations — but this is harder to observe through a shared branch predictor.

I would prefer code that's not obfuscated. We sometimes have code that's obfuscated because it has to be constant-time (see `constant_time.c`), but it comes with intermediate variables and comments explaining what's going on. Not one giant expression with four variables where you have to carefully examine every possibility to tell what's going on. The existing comment was adequate to explain the weirdness in the old code, it is not sufficient for the new code.

gstrauss replied:
Your statements about the `for()` loops are true, but are not touched by this PR.
This PR removes some other branches.  Incremental improvement.

In a separate PR from this one, If you wanted to *improve* the `for()` loops to better hide `A->n`, you could
```
    k = 0;
    for( i = A->n; i > 0; i-- )
        if( ( A->p[i - 1] != 0 ) & !k )
            k = i;
    i = k;
```
Full timing safety would require more invasive changes.

gstrauss replied:
I made some changes to continue using `result_is_zero`, but with fewer branches.
:rocket:

gstrauss commented:
I have added an additional commit with changes requested by @gilles-peskine-arm 
Please squash the commits before merge.
